numbat/src/typechecker.rs

```diff
@@ -487,134 +517,182 @@ impl TypeChecker {
-                                },
-                                span_expected: lhs.full_span(),
-                                expected_name: " left hand side",
-                                expected_dimensions: self
-                                    .registry
-                                    .get_derived_entry_names_for(&lhs_type),
-                                expected_type: lhs_type,
-                                span_actual: rhs.full_span(),
-                                actual_name: "right hand side",
-                                actual_name_for_fix: "expression on the right hand side",
-                                actual_dimensions: self
-                                    .registry
-                                    .get_derived_entry_names_for(&rhs_type),
-                                actual_type: rhs_type,
-                            },
-                        ))
+                // DateTime types need special handling here, since they're not scalars with dimensions, yet some select binary operators can be applied to them
+                // TODO how to better handle all the operations we want to support with date
+                if lhs_checked.get_type() == Type::DateTime {
+                    let rhs_is_time = dtype(&rhs_checked)
+                        .ok()
+                        .map(|t| t.is_time_dimension(&self.registry))
+                        .unwrap_or(false);
+                    let rhs_is_datetime = rhs_checked.get_type() == Type::DateTime;
+
+                    if *op == BinaryOperator::Sub && rhs_is_datetime {
+                        // TODO error handling
+                        let time = self
+                            .registry
+                            .get_base_representation_for_name("Time")
+                            .unwrap();
+
+                        typed_ast::Expression::BinaryOperatorForDate(
+                            *span_op,
+                            *op,
+                            Box::new(lhs_checked),
+                            Box::new(rhs_checked),
+                            Type::Dimension(time),
+                            true,
+                        )
+                    } else if *op == BinaryOperator::Add
+                        || *op == BinaryOperator::Sub && rhs_is_time
+                    {
+                        typed_ast::Expression::BinaryOperatorForDate(
+                            *span_op,
+                            *op,
+                            Box::new(lhs_checked),
+                            Box::new(rhs_checked),
+                            Type::DateTime,
+                            false,
+                        )
                     } else {
-                        Ok(Type::Dimension(lhs_type))
+                        return Err(TypeCheckError::IncompatibleType(IncompatibleTypeError {
+                            expected_name: "a unit of Time",
```

Comment:
This usually contains a dimension name, not a unit. This error message is also currently shown when doing something stupid like `now() -> seconds`, which is confusing.

Reply:
Yeah, the error handling and error formatting needs some attention.  The new `IncompatibleTypeError` type started as a copy/paste from `IncompatibleDimensionsError` and so it contains some vestigial fields that might not be needed, or might not be named/used correctly